pkg/eval/unix/umask.go

```diff
@@ -0,0 +1,85 @@
+// +build !windows,!plan9,!js
+
+package unix
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/elves/elvish/pkg/eval/vars"
+)
+
+//elvdoc:var umask
+//
+// The file mode creation mask. Its value is a string in Elvish octal
+// representation; e.g. 0o027. This makes it possible to use it in any context
+// that expects a `$number`.
+//
+// When assigning a new value a string is implicitly treated as an octal
+// number. If that fails the usual rules for interpreting
+// [numbers](./language.html#number) are used. The following arer equivalent:
+// `unix:umask = 027` and `unix:umask = 0o27`. You can also assign to it a
+// `float64` data type that has no fractional component.
+// The assigned value must be within the range [0 ... 0o777], otherwise the
+// assignment will throw an exception.
+//
+// You can do a temporary assignment to affect a single command; e.g.
+// `umask=077 touch a_file`. After the command completes the old umask will be
+// restored. **Warning**: Since the umask applies to the entire process, not
+// individual threads, changing it temporarily in this manner is dangerous if
+// you are doing anything in parallel. Such as via the
+// [`peach`](ref/builtin.html#peach) command.
+
+// UmaskVariable is a variable whose value always reflects the current file
+// creation permission mask. Setting it changes the current file creation
+// permission mask for the process (not an individual thread).
+type UmaskVariable struct{}
+
+var _ vars.Var = UmaskVariable{}
+
+// Get returns the current file creation umask as a string.
+func (UmaskVariable) Get() interface{} {
+	// Note: The seemingly redundant syscall is because the unix.Umask() API
+	// doesn't allow querying the current value without changing it. So ensure
+	// we reinstate the curent value.
+	umask := unix.Umask(0)
```

Comment:
Re adding a mutex: since Elvish does not use Umask elsewhere concurrently, guarding UmaskVariable with a mutex will prevent possible race conditions. Elvish does not link with arbitrary Go code, so as long as we control the Elvish source code we have a guarantee.

The fact that Fish has a race condition does not make the race condition less likely :)

Reply:
I still think a mutex for this is overkill but it's trivial to do and won't hurt performance unless someone is intentionally trying to trigger the race -- in which case the mutex is needed -- so I'll go ahead and add one :-)

Reply:
I added a mutex: `var umaskMutex sync.Mutex`. Is something wrong with how I used it?

Reply:
Ah, I missed that, sorry.

It's correct, but a small nitpick: use a `RWMutex` to allow concurrent reads. Not that it's likely to have a lot of concurrent reads, more of a stylistic issue.